File: codeforces/898/H.py
import sys
lines = list(map(str.strip, sys.stdin.readlines()))
from collections import deque
sys.setrecursionlimit(300000)
# TODO Remember to add int wrapping if using dict

ii = 1
while ii < len(lines):
    n, marcel, valeriu = map(int, lines[ii].split())
    marcel -= 1
    valeriu -= 1
    ii += 1
    adj_lists = {i: [] for i in range(n)}
    for _ in range(n):
        fr, to = map(int, lines[ii].split())
        adj_lists[fr-1].append(to-1)
        adj_lists[to-1].append(fr-1)
        ii+=1
    visited = set()
    loop_starts = set()
    def dfs(prev, i):
        in_loop = False
        visited.add(i)
        for nbr in adj_lists[i]:
            if nbr == prev:
                continue
            if nbr in visited:
                in_loop = True
            else:
                in_loop |= dfs(i, nbr)
        if in_loop:
            loop_starts.add(i)
        return in_loop
    dfs(-1, marcel)
    def dist_to_loops(start):
        result = {}
        q = deque([start])
        visited = set()
        d = 0
        while q:
            for _ in range(len(q)):
                node = q.popleft()
                if node in visited:
                    continue
                visited.add(node)
                if node in loop_starts:
                    result[node] = d
                for nbr in adj_lists[node]:
                    if nbr not in visited:
                        q.append(nbr)
            d += 1
        return result
    marcel_dist = dist_to_loops(marcel)
    valeriu_dist = dist_to_loops(valeriu)
    # print(marcel_dist)
    # print(valeriu_dist)
    for k, v in valeriu_dist.items():
        if v < marcel_dist[k]:
            print('YES')
            break
    else:
        print('NO')
    # print()

File: codeforces/898/H.cpp
#include  <bits/stdc++.h>
#include <string>
#include <vector>
#include <iostream>
#include <math.h>
#include <algorithm>

using namespace std;

#define ll long long

template <class T>
void print_v(vector<T> &v) { cout << "["; for (auto x : v) cout << x << ","; cout << "\b]\n"; }

ll min(ll a,int b) { if (a<b) return a; return b; }
ll min(int a,ll b) { if (a<b) return a; return b; }
ll max(ll a,int b) { if (a>b) return a; return b; }
ll max(int a,ll b) { if (a>b) return a; return b; }
ll gcd(ll a,ll b) { if (b==0) return a; return gcd(b, a%b); }
ll lcm(ll a,ll b) { return a/gcd(a,b)*b; }
string to_upper(string a) { for (int i=0;i<(int)a.size();++i) if (a[i]>='a' && a[i]<='z') a[i]-='a'-'A'; return a; }
string to_lower(string a) { for (int i=0;i<(int)a.size();++i) if (a[i]>='A' && a[i]<='Z') a[i]+='a'-'A'; return a; }
bool prime(ll a) { if (a==1) return 0; for (int i=2;i<=round(sqrt(a));++i) if (a%i==0) return 0; return 1; }


int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        int n, marcel, valeriu;
        cin >> n >> marcel >> valeriu;
        marcel--;
        valeriu--;

        vector<vector<int>> adj_lists(n);
        for (int i = 0; i < n; i++) {
            int fr, to;
            cin >> fr >> to;
            adj_lists[fr-1].push_back(to-1);
            adj_lists[to-1].push_back(fr-1);
        }

        unordered_set<int> visited;
        unordered_set<int> loop_starts;

        function<bool(int, int)> dfs = [&](int prev, int i) {
            bool in_loop = false;
            visited.insert(i);
            for (int nbr : adj_lists[i]) {
                if (nbr == prev) continue;
                if (visited.find(nbr) != visited.end()) {
                    in_loop = true;
                } else {
                    in_loop |= dfs(i, nbr);
                }
            }
            if (in_loop) {
                loop_starts.insert(i);
            }
            return in_loop;
        };

        dfs(-1, marcel);

        auto dist_to_loops = [&](int start) {
            unordered_map<int, int> result;
            deque<int> q = {start};
            unordered_set<int> visited;

            int d = 0;
            while (!q.empty()) {
                int len = q.size();
                for (int i = 0; i < len; i++) {
                    int node = q.front();
                    q.pop_front();
                    if (visited.find(node) != visited.end()) continue;
                    visited.insert(node);
                    if (loop_starts.find(node) != loop_starts.end()) {
                        result[node] = d;
                    }
                    for (int nbr : adj_lists[node]) {
                        if (visited.find(nbr) == visited.end()) {
                            q.push_back(nbr);
                        }
                    }
                }
                d++;
            }
            return result;
        };

        auto marcel_dist = dist_to_loops(marcel);
        auto valeriu_dist = dist_to_loops(valeriu);

        bool found = false;
        for (const auto& pair : valeriu_dist) {
            int k = pair.first;
            int v = pair.second;
            if (v < marcel_dist[k]) {
                cout << "YES\n";
                found = true;
                break;
            }
        }
        if (!found) {
            cout << "NO\n";
        }
    }

    return 0;
}
